diff -Nura iw-5.9/Makefile iw-5.9_new/Makefile
--- iw-5.9/Makefile	2020-10-02 18:18:41.000000000 +0800
+++ iw-5.9_new/Makefile	2021-05-28 15:52:00.246697746 +0800
@@ -18,6 +18,11 @@
 CFLAGS += -Werror-implicit-function-declaration -Wsign-compare -Wno-unused-parameter
 CFLAGS += -Wdeclaration-after-statement
 CFLAGS += $(CFLAGS_EVAL)
+CFLAGS += -I$(abspath ../../libnl/libnl-3.2.25/include/)
+LDFLAGS += -L$(abspath ../../libnl/libnl-3.2.25/lib/.libs/)
+CFLAGS += -DCONFIG_LIBNL30
+LIBS += -lnl-genl-3 -lnl-3
+NLLIBNAME = libnl-3.0
 
 _OBJS := $(sort $(patsubst %.c,%.o,$(wildcard *.c)))
 VERSION_OBJS := $(filter-out version.o, $(_OBJS))
@@ -25,6 +30,8 @@
 
 ALL = iw
 
+NO_PKG_CONFIG = 1
+
 ifeq ($(NO_PKG_CONFIG),)
 NL3xFOUND := $(shell $(PKG_CONFIG) --atleast-version=3.2 libnl-3.0 && echo Y)
 ifneq ($(NL3xFOUND),Y)
