--- a/drivers/net/wireless/mtk/mt7628/mt7628_wifi/embedded/include/os/rt_linux_cmm.h
+++ b/drivers/net/wireless/mtk/mt7628/mt7628_wifi/embedded/include/os/rt_linux_cmm.h
@@ -454,8 +454,8 @@ extern RTMP_USB_CONFIG *pRtmpUsbConfig;
 #ifdef BB_SOC
 #define RX_RING_SIZE            64 
 #else
-//#define RX_RING_SIZE            128 /*64 */
-#define RX_RING_SIZE            64 /*64 */
+#define RX_RING_SIZE            128 /*64 */
+//#define RX_RING_SIZE            64 /*64 */
 #endif
 #ifdef BB_SOC
 #define MGMT_RING_SIZE          64
--- a/drivers/net/wireless/mtk/mt7628/mt7628_wifi/embedded/include/rtmp_def.h
+++ b/drivers/net/wireless/mtk/mt7628/mt7628_wifi/embedded/include/rtmp_def.h
@@ -118,14 +118,14 @@
 	1600; Or kernel will crash in deaggregate_AMSDU_announce() for EAPOL packet
 	in enterprise WPA mode.
 */
-#define MGMT_DMA_BUFFER_SIZE    1600	/*2048 */
+#define MGMT_DMA_BUFFER_SIZE    2048	/*2048 */
 
-#define RX_BUFFER_AGGRESIZE     3840	/*3904 //3968 //4096 //2048 //4096 */
-#define RX1_BUFFER_SIZE         1700    /* 1700 //512 */
-#define RX_BUFFER_NORMSIZE      3840	/*3904 //3968 //4096 //2048 //4096 */
+#define RX_BUFFER_AGGRESIZE     2048	/*3904 //3968 //4096 //2048 //4096 */
+#define RX1_BUFFER_SIZE         2048    /* 1700 //512 */
+#define RX_BUFFER_NORMSIZE      2048	/*3904 //3968 //4096 //2048 //4096 */
 #define TX_BUFFER_NORMSIZE		RX_BUFFER_NORMSIZE
 #define MAX_FRAME_SIZE          2346	/* Maximum 802.11 frame size */
-#define MAX_AGGREGATION_SIZE    3840	/*3904 //3968 //4096 */
+#define MAX_AGGREGATION_SIZE    2048	/*3904 //3968 //4096 */
 #define MAX_NUM_OF_TUPLE_CACHE  2
 #define MAX_MCAST_LIST_SIZE     32
 #define MAX_LEN_OF_VENDOR_DESC  64
