Index: linux-3.10.108/arch/mips/ralink/Kconfig
===================================================================
--- linux-3.10.108.orig/arch/mips/ralink/Kconfig
+++ linux-3.10.108/arch/mips/ralink/Kconfig
@@ -443,4 +443,9 @@ config  RALINK_SYSTICK
 config  UBOOT_CMDLINE
 	bool "UBOOT_CMDLINE"
 	depends on RALINK_MT7620 || RALINK_MT7621 || RALINK_MT7628
+	default n
+
+config NODEBUG
+	bool "NO DEBUG"
+	depends on RALINK_MT7620 || RALINK_MT7621 || RALINK_MT7628
 	default n
\ No newline at end of file
Index: linux-3.10.108/kernel/printk.c
===================================================================
--- linux-3.10.108.orig/kernel/printk.c
+++ linux-3.10.108/kernel/printk.c
@@ -51,12 +51,18 @@
 #define CREATE_TRACE_POINTS
 #include <trace/events/printk.h>
 
-/* printk's without a loglevel use this.. */
-#define DEFAULT_MESSAGE_LOGLEVEL CONFIG_DEFAULT_MESSAGE_LOGLEVEL
-
 /* We show everything that is MORE important than this.. */
 #define MINIMUM_CONSOLE_LOGLEVEL 1 /* Minimum loglevel we let people use */
+
+/* printk's without a loglevel use this.. */
+#ifndef CONFIG_NODEBUG
+#define DEFAULT_MESSAGE_LOGLEVEL CONFIG_DEFAULT_MESSAGE_LOGLEVEL
 #define DEFAULT_CONSOLE_LOGLEVEL 7 /* anything MORE serious than KERN_DEBUG */
+#else
+#define DEFAULT_MESSAGE_LOGLEVEL 1
+#define DEFAULT_CONSOLE_LOGLEVEL 1 /* anything MORE serious than KERN_DEBUG */
+#endif
+
 
 int console_printk[4] = {
 	DEFAULT_CONSOLE_LOGLEVEL,	/* console_loglevel */
@@ -1831,6 +1837,7 @@ static int __init console_setup(char *st
 
 	__add_preferred_console(buf, idx, options, brl_options);
 	console_set_on_cmdline = 1;
+
 	return 1;
 }
 __setup("console=", console_setup);
Index: linux-3.10.108/arch/mips/ralink/printf.c
===================================================================
--- linux-3.10.108.orig/arch/mips/ralink/printf.c
+++ linux-3.10.108/arch/mips/ralink/printf.c
@@ -107,6 +107,9 @@ void prom_printf(char *fmt, ...)
 	unsigned long flags;
 
 	int putPromChar(char);
+	#ifdef CONFIG_NODEBUG
+	return;
+	#endif
 
 	spin_lock_irqsave(&con_lock, flags);
 	va_start(args, fmt);
Index: linux-3.10.108/drivers/tty/serial/8250/8250_core.c
===================================================================
--- linux-3.10.108.orig/drivers/tty/serial/8250/8250_core.c
+++ linux-3.10.108/drivers/tty/serial/8250/8250_core.c
@@ -2955,6 +2955,7 @@ static int __init serial8250_console_ini
 	register_console(&serial8250_console);
 	return 0;
 }
+
 console_initcall(serial8250_console_init);
 
 int serial8250_find_port(struct uart_port *p)
@@ -3402,9 +3403,11 @@ static void __exit serial8250_exit(void)
 #endif
 }
 
+
 module_init(serial8250_init);
 module_exit(serial8250_exit);
 
+
 EXPORT_SYMBOL(serial8250_suspend_port);
 EXPORT_SYMBOL(serial8250_resume_port);
 
